File: packages/prebid_mobile_sdk_admob/ios/prebid_mobile_sdk_admob/Sources/prebid_mobile_sdk_admob/AdMobBannerAdViewFactory.swift
import Flutter
import UIKit
import GoogleMobileAds
import PrebidMobile
import PrebidMobileAdMobAdapters

/// PlatformView factory for AdMob-mediated banners. The rendered view is the
/// Google Mobile Ads `BannerView`; Prebid's `MediationBannerAdUnit` runs the
/// auction and passes the winning bid to AdMob via the Prebid AdMob adapter.
class AdMobBannerAdViewFactory: NSObject, FlutterPlatformViewFactory {

    private let messenger: FlutterBinaryMessenger

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return AdMobBannerPlatformView(
            frame: frame,
            viewId: viewId,
            messenger: messenger,
            args: args as? [String: Any] ?? [:]
        )
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class AdMobBannerPlatformView: NSObject, FlutterPlatformView, GoogleMobileAds.BannerViewDelegate {

    private let gadBanner: GoogleMobileAds.BannerView
    private let methodChannel: FlutterMethodChannel
    private let adSize: CGSize

    // Retained for the lifetime of the view — the auction runs through these.
    private var mediationDelegate: AdMobMediationBannerUtils?
    private var mediationAdUnit: MediationBannerAdUnit?

    init(
        frame: CGRect,
        viewId: Int64,
        messenger: FlutterBinaryMessenger,
        args: [String: Any]
    ) {
        let configId = args["configId"] as? String ?? ""
        let adMobAdUnitId = args["adMobAdUnitId"] as? String ?? ""
        let width = args["width"] as? Int ?? 320
        let height = args["height"] as? Int ?? 50
        let autoLoad = args["autoLoad"] as? Bool ?? true

        adSize = CGSize(width: width, height: height)

        methodChannel = FlutterMethodChannel(
            name: "prebid_mobile_sdk_admob/banner_\(viewId)",
            binaryMessenger: messenger
        )

        // 1. Create the GMA request and banner view.
        let gadRequest = Request()
        gadBanner = GoogleMobileAds.BannerView(adSize: adSizeFor(cgSize: adSize))
        gadBanner.adUnitID = adMobAdUnitId

        super.init()

        gadBanner.delegate = self
        gadBanner.rootViewController = UIApplication.shared.keyWindow?.rootViewController

        // 2. Prebid mediation utils + ad unit.
        let mediationDelegate = AdMobMediationBannerUtils(gadRequest: gadRequest, bannerView: gadBanner)
        self.mediationDelegate = mediationDelegate
        let adUnit = MediationBannerAdUnit(
            configID: configId,
            size: adSize,
            mediationDelegate: mediationDelegate
        )
        mediationAdUnit = adUnit

        if autoLoad {
            // 3. Fetch demand, then let AdMob run its waterfall and render.
            adUnit.fetchDemand { [weak self] _ in
                self?.gadBanner.load(gadRequest)
            }
        }
    }

    func view() -> UIView {
        return gadBanner
    }

    // MARK: - BannerViewDelegate

    func bannerViewDidReceiveAd(_ bannerView: GoogleMobileAds.BannerView) {
        methodChannel.invokeMethod("onAdSize", arguments: [
            "width": Double(adSize.width),
            "height": Double(adSize.height),
        ])
        methodChannel.invokeMethod("onAdLoaded", arguments: nil)
        methodChannel.invokeMethod("onAdDisplayed", arguments: nil)
    }

    func bannerView(
        _ bannerView: GoogleMobileAds.BannerView,
        didFailToReceiveAdWithError error: Error
    ) {
        mediationAdUnit?.adObjectDidFailToLoadAd(adObject: gadBanner, with: error)
        methodChannel.invokeMethod("onAdFailed", arguments: error.localizedDescription)
    }

    func bannerViewDidRecordClick(_ bannerView: GoogleMobileAds.BannerView) {
        methodChannel.invokeMethod("onAdClicked", arguments: nil)
    }

    func bannerViewDidDismissScreen(_ bannerView: GoogleMobileAds.BannerView) {
        methodChannel.invokeMethod("onAdClosed", arguments: nil)
    }
}

File: packages/prebid_mobile_sdk_admob/android/src/main/kotlin/com/prebid/prebid_mobile_sdk_admob/AdMobBannerAdViewFactory.kt
package com.prebid.prebid_mobile_sdk_admob

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.View
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize as GmaAdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.LoadAdError
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import org.prebid.mobile.AdSize
import org.prebid.mobile.admob.AdMobMediationBannerUtils
import org.prebid.mobile.admob.PrebidBannerAdapter
import org.prebid.mobile.api.mediation.MediationBannerAdUnit

/// PlatformView factory for AdMob-mediated banners. The rendered view is the
/// Google Mobile Ads [AdView]; Prebid's [MediationBannerAdUnit] runs the auction
/// and passes the winning bid to AdMob via the Prebid AdMob adapter.
class AdMobBannerAdViewFactory(
    private val messenger: BinaryMessenger,
    private val activityProvider: () -> Activity?,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val params = args as? Map<*, *> ?: emptyMap<String, Any>()
        return AdMobBannerPlatformView(
            activityProvider() ?: context,
            viewId,
            messenger,
            params,
        )
    }
}

class AdMobBannerPlatformView(
    context: Context,
    viewId: Int,
    messenger: BinaryMessenger,
    params: Map<*, *>,
) : PlatformView {

    private val adView: AdView = AdView(context)
    private val methodChannel: MethodChannel
    private var adUnit: MediationBannerAdUnit? = null

    init {
        val configId = params["configId"] as? String ?: ""
        val adMobAdUnitId = params["adMobAdUnitId"] as? String ?: ""
        val width = params["width"] as? Int ?: 320
        val height = params["height"] as? Int ?: 50
        val autoLoad = params["autoLoad"] as? Boolean ?: true

        methodChannel = MethodChannel(messenger, "prebid_mobile_sdk_admob/banner_$viewId")

        adView.setAdSize(GmaAdSize(width, height))
        adView.adUnitId = adMobAdUnitId
        adView.adListener = object : AdListener() {
            override fun onAdLoaded() {
                methodChannel.invokeMethod(
                    "onAdSize",
                    mapOf("width" to width.toDouble(), "height" to height.toDouble()),
                )
                methodChannel.invokeMethod("onAdLoaded", null)
                methodChannel.invokeMethod("onAdDisplayed", null)
            }

            override fun onAdFailedToLoad(error: LoadAdError) {
                methodChannel.invokeMethod("onAdFailed", error.message)
            }

            override fun onAdClicked() {
                methodChannel.invokeMethod("onAdClicked", null)
            }

            override fun onAdClosed() {
                methodChannel.invokeMethod("onAdClosed", null)
            }
        }

        // Prebid targeting keywords are written into this Bundle by the adapter.
        val extras = Bundle()
        val request = AdRequest.Builder()
            .addNetworkExtrasBundle(PrebidBannerAdapter::class.java, extras)
            .build()

        val mediationUtils = AdMobMediationBannerUtils(extras, adView)
        adUnit = MediationBannerAdUnit(
            context,
            configId,
            AdSize(width, height),
            mediationUtils,
        )

        if (autoLoad) {
            adUnit?.fetchDemand {
                // The bid (if any) is now attached to the request extras; let
                // AdMob run its waterfall and render.
                adView.loadAd(request)
            }
        }
    }

    override fun getView(): View = adView

    override fun dispose() {
        adUnit?.destroy()
        adUnit = null
        adView.destroy()
    }
}
